### PR TITLE
Handle stuck queued tasks in Celery for db backend

### LINKED ISSUE #19699: task_instances stuck in "queued" and are missing corresponding celery_taskmeta entries
### Apache Airflow version

2.2.2 (latest released)

### Operating System

Linux Mint 20.2

### Versions of Apache Airflow Providers

apache-airflow-providers-celery = "2.1.0"
apache-airflow-providers-papermill = "^2.1.0"
apache-airflow-providers-postgres = "^2.2.0"
apache-airflow-providers-google = "^6.1.0"

### Deployment

Docker-Compose

### Deployment details

Docker-compose deploys into our GCP k8s cluster

### What happened

Hi,

we're running Airflow for our ETL pipelines. 

Our DAGs run in parallel and we spawn a fair bit of parallel DAGs and tasks every morning for our pipelines.
We run our Airflow in a k8s cluster in GCP and we use Celery for our executors.
And we use autopilot to dynamically scale up and down the cluster as the workload increases or decreases, thereby sometimes tearing down airflow workers.

Ever since upgrading to Airflow 2.0 we've had a lot of problems with tasks getting stuck in "queued" or "running", and we've had to clean it up by manually failing the stuck tasks and re-running the DAGs.
Following the discussions here over the last months it looks we've not been alone :-)

But, after upgrading to Airflow 2.2.1 we saw a significant decrease in the number of tasks getting stuck (yay!), something we hoped for given the bug fixes for the scheduler in that release.
However, we still have a few tasks getting stuck (Stuck = "Task in queued") on most mornings that require the same manual intervention. 

I've started digging in the Airflow DB trying to see where there's a discrepancy, and every time a task gets stuck it's missing a correspondning task in the table "celery_taskmeta".
This is a consistent pattern for the tasks that are stuck with us at this point. The task has rows in the tables "task_instance", "job", and "dag_run" with IDs referencing each other. 

But the "external_executor_id" in "task_instance" is missing a corresponding entry in the "celery_taskmeta" table. So nothing ever gets executed and the task_instance is forever stuck in "queued" and never cleaned up by the scheduler.

I can see in "dag_run::last_scheduling_decision" that the scheduler is continuously re-evaluating this task as the timestamp is updated, so it's inspecting it at least, but it leaves everything in the "queued" state. 

The other day I bumped our Airflow to 2.2.2, but we still get the same behavior.
And finally, whenever we get tasks that are stuck in "Queued" in this way they usually occur within the same few seconds timestamp-wise, and it correlates timewise to a timestamp when autopilot scaled down the number of airflow-workers.

If the tasks end up in this orphaned/queued state then they never get executed and are stuck until we fail them. Longest I've seen so far is a few days in this state until the task was discovered.
Restarting the scheduler does not resolve this issue and tasks are still stuck in "queued" afterwards.

Would it be possible (and a good idea?) to include in the scheduler a check if a "task_instance" row has a corresponding row in the "celery_taskmeta", and if its missing in "celery_taskmeta" after a given amount of time clean it up?
After reading about and watching Ash Berlin-Taylor's most excellent video on a deep dive into the Airflow's scheduler this does seem exactly like the check that we should add to the scheduler?

Also if there's any data I can dig out and provide for this, don't hesitate to let me know.

### What you expected to happen

I expect orphaned tasks in the state queued and that are missing a corresponding entry in celery_taskmeta to be cleaned up and re-executed by the scheduler.

### How to reproduce

Currently no deterministic way to reproduce other than a large amount of tasks and then remove a worker at just the right time.
Occurs every morning in a handful of tasks, but no deterministic way to reproduce it.

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ephraimbuddy**
Move the state of stuck queued tasks in Celery to Scheduled so that
the Scheduler can queue them again
Closes: https://github.com/apache/airflow/issues/19699

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/executors/celery_executor.py [hunk 1]
```diff
@@ -337,6 +344,12 @@ def sync(self) -> None:
 
         if self.adopted_task_timeouts:
             self._check_for_stalled_adopted_tasks()
+        # Use EventScheduler to avoid checking this too often
+        self.event_scheduler = EventScheduler()
+        self.event_scheduler.call_regular_interval(
+            self.stuck_queued_task_check_interval,
+            self._clear_stuck_queued_tasks(),
+        )
```
Comment (ashb) [inline airflow/executors/celery_executor.py]: This isn't quite right -- it will re-create a new "timer" each time `sync()` is called, but never check it -- for that we have the instance live longer than one sync() call and to call `timers.run(blocking=False)`

File: airflow/executors/celery_executor.py [hunk 2]
```diff
@@ -377,6 +390,36 @@ def _check_for_stalled_adopted_tasks(self):
...
+
+            self.log.info("Checking task %s", task)
+
+            if task.external_executor_id:
+                continue
+            if task.key in self.queued_tasks or task.key in self.running:
+                continue
+            # We use the task_adoption_timeout to decide if the task is stuck
+            if task.queued_dttm > utcnow() - datetime.timedelta(seconds=self.task_adoption_timeout):
+                continue
```
Comment (ashb) [inline airflow/executors/celery_executor.py]: You can do most of these at the SQL level and have the query only return the rows that should be cleared.

But we probably need to _only_ do tasks owned by "this" executor.

File: airflow/executors/celery_executor.py (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/executors/celery_executor.py]: I will create the timer at `start` method. Thanks!

File: airflow/executors/celery_executor.py (hunk 2 above)
Comment (ephraimbuddy) [reply airflow/executors/celery_executor.py]: Since the external_executor_id of the task instance is no longer available in celery is there a way I could get only for the tasks in this executor? I'm currently searching broadly using AsyncResult to see if another executor is running this task

File: airflow/executors/celery_executor.py [hunk 3]
```diff
@@ -377,6 +387,43 @@ def _check_for_stalled_adopted_tasks(self):
...
+                self.log.info(
+                    'TaskInstance: %s found in queued state for more than %s seconds, rescheduling',
+                    task,
+                    self.task_adoption_timeout.total_seconds(),
+                )
+                session.query(TaskInstance).filter(
+                    TaskInstance.dag_id == task.dag_id,
+                    TaskInstance.task_id == task.task_id,
+                    TaskInstance.run_id == task.run_id,
+                ).update({TaskInstance.state: State.SCHEDULED})
```
Comment (uranusjr) [inline airflow/executors/celery_executor.py]: Why not `task.state = State.SCHEDULED`?

File: airflow/executors/celery_executor.py [hunk 4]
```diff
@@ -377,6 +387,43 @@ def _check_for_stalled_adopted_tasks(self):
...
+        For some reasons, tasks can get stuck in queued state in DB while still not in
+        self.queued_tasks and not in self.running_tasks.
+
+        In such situation, we update the task instance state to scheduled so that
+        it can be queued again. We chose to use task_adoption_timeout to decide
+        """
+        self.log.debug("Checking for stuck queued tasks")
+        queued_tasks = session.query(TaskInstance).filter(TaskInstance.state == State.QUEUED).all()
+
+        for task in queued_tasks:
```
Comment (uranusjr) [inline airflow/executors/celery_executor.py]:
```suggestion
        for task in session.query(TaskInstance).filter(TaskInstance.state == State.QUEUED):
```

File: airflow/executors/celery_executor.py (hunk 3 above)
Comment (ephraimbuddy) [reply airflow/executors/celery_executor.py]: I will try that, copied code :)

File: airflow/executors/celery_executor.py [hunk 5]
```diff
@@ -377,6 +387,39 @@ def _check_for_stalled_adopted_tasks(self):
             for key in timedout_keys:
                 self.change_state(key, State.FAILED)
 
+    @provide_session
+    def _clear_stuck_queued_tasks(self, session=None):
+        """
+        For some reasons, tasks can get stuck in queued state in DB while still not in
+        self.queued_tasks and not in self.running_tasks.
```
Comment (tanelk) [inline airflow/executors/celery_executor.py]: Could you give here a little explanation, why tasks can get stuck. From the "For some reasons..." it sounds like there is some bug and this here is just a bandage.

Comment (ephraimbuddy) [reply airflow/executors/celery_executor.py]: Updated

File: tests/executors/test_celery_executor.py [hunk 6]
```diff
@@ -411,6 +404,107 @@ def test_check_for_stalled_adopted_tasks_goes_in_ordered_fashion(self):
...
+        # Since AsyncResult is mocked, the returned_value is not PENDING
+        # We end up not clearing the task
+        assert ti.state == State.QUEUED
+
+    @pytest.mark.backend("mysql", "postgres")
+    @pytest.mark.parametrize(
+        "last_check_time, state",
+        [
+            (time.time() - 400, State.SCHEDULED),
+            (time.time() - 200, State.QUEUED),
```
Comment (uranusjr) [inline tests/executors/test_celery_executor.py]: Consider using `freezegun` here—calculating this against actual time may cause issues if the test suite is running slowly, since these `time.time()` are evaluated at import time, but the test may not actually start immediately after import (if the test runner decides to run other tests first).

**Review by uranusjr — APPROVED** [review]

**Comment by potiuk** [comment]
There are some sqlite tests failing here - likely related.

File: airflow/executors/celery_executor.py [hunk 7]
```diff
@@ -377,6 +384,52 @@ def _check_for_stalled_adopted_tasks(self):
...
+            celery_task_ids = [t.task_id for t in session_.query(task_cls.task_id).all()]
+        self.log.debug("Checking for stuck queued tasks")
+
+        max_allowed_time = utcnow() - self.task_adoption_timeout
+
+        for task in session.query(TaskInstance).filter(
+            TaskInstance.state == State.QUEUED, TaskInstance.queued_dttm < max_allowed_time
+        ):
+
+            self.log.info("Checking task %s", task)
```
Comment (kaxil) [inline airflow/executors/celery_executor.py]:
```suggestion
```

File: airflow/executors/celery_executor.py [hunk 8]
```diff
@@ -377,6 +384,52 @@ def _check_for_stalled_adopted_tasks(self):
             for key in timedout_keys:
                 self.change_state(key, State.FAILED)
 
+    @provide_session
+    def _clear_stuck_queued_tasks(self, session=None):
```
Comment (kaxil) [inline airflow/executors/celery_executor.py]:
```suggestion
    def _clear_stuck_queued_tasks(self: Session = NEW_SESSION) -> None:
```

while we are at it and since it is a new method

File: tests/executors/test_celery_executor.py [hunk 9]
```diff
@@ -411,6 +406,171 @@ def test_check_for_stalled_adopted_tasks_goes_in_ordered_fashion(self):
...
+        ti = create_task_instance(state=State.QUEUED)
+        ti.queued_dttm = timezone.utcnow() - timedelta(days=2)
+        ti.external_executor_id = '231'
+        session.merge(ti)
+        session.flush()
+        executor = celery_executor.CeleryExecutor()
+        executor.queued_tasks = {ti.key: AsyncResult("231")}
+        executor._clear_stuck_queued_tasks()
+        session.flush()
+        ti = session.query(TaskInstance).filter(TaskInstance.task_id == ti.task_id).one()
```
Comment (kaxil) [inline tests/executors/test_celery_executor.py:447]: based on the test name, you should check the following:

```
assert executor.queued_tasks == {ti.key: AsyncResult("231")}
```

File: tests/executors/test_celery_executor.py [hunk 10]
```diff
@@ -411,6 +406,171 @@ def test_check_for_stalled_adopted_tasks_goes_in_ordered_fashion(self):
...
+        ti = create_task_instance(state=State.QUEUED)
+        ti.queued_dttm = timezone.utcnow() - timedelta(days=2)
+        ti.external_executor_id = '231'
+        session.merge(ti)
+        session.flush()
+        executor = celery_executor.CeleryExecutor()
+        executor.running = {ti.key: AsyncResult("231")}
+        executor._clear_stuck_queued_tasks()
+        session.flush()
+        ti = session.query(TaskInstance).filter(TaskInstance.task_id == ti.task_id).one()
```
Comment (kaxil) [inline tests/executors/test_celery_executor.py:465]:
```
assert executor.running == {ti.key: AsyncResult("231")}
```

**Review by kaxil — APPROVED** [review]
Minor changes suggsted, lgtm otherwise

File: airflow/config_templates/config.yml [hunk 11]
```diff
@@ -1678,6 +1678,13 @@
       type: string
       example: ~
       default: "False"
+    - name: stuck_queued_task_check_interval
+      description: |
+        How often to check for stuck queued task (in seconds)
+      version_added: 2.2.3
```
Comment (kaxil) [inline airflow/config_templates/config.yml]:
```suggestion
      version_added: 2.3.0
```

**Review by kaxil — APPROVED** [review]

File: airflow/executors/celery_executor.py [hunk 12]
```diff
@@ -377,6 +385,49 @@ def _check_for_stalled_adopted_tasks(self):
...
 
+    @provide_session
+    def _clear_stuck_queued_tasks(self, session: Session = NEW_SESSION) -> None:
+        """
+        Tasks can get stuck in queued state in DB while still not in
+        worker. This happens when the worker is autoscaled down and
+        the task is queued but has not been picked up by any worker prior to the scaling.
+
+        In such situation, we update the task instance state to scheduled so that
+        it can be queued again. We chose to use task_adoption_timeout to decide
```
Comment (SamWheating) [inline airflow/executors/celery_executor.py]:
```suggestion
        it can be queued again. We chose to use task_adoption_timeout to decide when
        a queued task is considered stuck and should be reschelduled.
```

File: airflow/executors/celery_executor.py [hunk 13]
```diff
@@ -377,6 +385,49 @@ def _check_for_stalled_adopted_tasks(self):
...
+        """
+        if not isinstance(app.backend, DatabaseBackend):
+            # We only want to do this for database backends where
```
Comment (SamWheating) [inline airflow/executors/celery_executor.py:400]: Interesting - we have seen this same issue when using a Redis results backend - is there a good explanation for why this would only happen with DB backends?

Comment (ephraimbuddy) [reply airflow/executors/celery_executor.py:400]: I can't figure how it could be solved in Redis. Do you have a reproductive step for Redis just like in #19699

Comment (kaxil) [reply airflow/executors/celery_executor.py:399]: Oh yes, atleast in that case, the comment should be clear with a proper TODO or just solved in this PR too for all the other backends.

**Comment by kristoffern** [comment]
@ephraimbuddy I'm trying to test the patch locally, and I'm sure I'm doing something wrong, but after the cluster is deployed the scheduler crashes with the following error.

```
file "/home/airflow/.local/lib/python3.8/site-packages/airflow/executors/celery_executor.py", line 54, in <module>
    from airflow.utils.session import NEW_SESSION, provide_session
ImportError: cannot import name 'NEW_SESSION' from 'airflow.utils.session' (/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/session.py)
```

I've looked in this PR and can't find the NEW_SESSION in the files changed, and I also can't find it in Airflows main branch.
I don't think I have access to the "astronomer:fix-stuck-task" repo/branch so I can't provide more details unfortunately.

Am I missing something, or was something missedin the PR?

I'm downloading this patch and applying it in my Dockerfile, I can provide the details if you want to.

https://patch-diff.githubusercontent.com/raw/apache/airflow/pull/19769.patch

**Comment by ephraimbuddy** [comment]
@kristoffern The NEW_SESSION is a new addition that is not in 2.2.3, remove every reference to NEW_SESSION on the code and it should work.

Also, remove the test changes before applying the patch

**Comment by kristoffern** [comment]
@ephraimbuddy Thanks :+1: 
I already removed all the test_* files (not anything else right?), thanks for the help on the NEW_SESSION.
Is NEW_SESSION part of 2.2.4 and 2.3.0 then? 

I'll update here with a comment if you want to after I tested the patch :+1:

**Comment by ephraimbuddy** [comment]
> Is NEW_SESSION part of 2.2.4 and 2.3.0 then?

Yeah, will be.

> I'll update here with a comment if you want to after I tested the patch 👍

I will appreciate that

**Comment by kristoffern** [comment]
@ephraimbuddy I've now deployed the modified patch-file to our Kubernetes cluster and tested since yesterday.

In short I started all our parallel DAG runs for the data processing/ETL that we do.
In the past it's reproduced the bug on virtually every attempt with tasks that get stuck in "Queued".

At the moment I've ran it 4 times and on none of the runs have I seen the bug occur. 
Now of course this isn't proof the bug is resolved, but it has drastically reduced the odds that the bug is still there :+1: 

So to summarize:

A: Patch doesn't break the system
B: Patch by all accounts fixes the stuck in queued bug to occur.

[19769-without-NEW_SESSION.patch.txt](https://github.com/apache/airflow/files/7977405/19769-without-NEW_SESSION.patch.txt)
.

Thank you for the hard work, and likely I will use this patch and modified Dockerfile in our deployment scripts until 2.2.4 or 2.3.0 is released :+1:

**Comment by kristoffern** [comment]
And finally (hopefully) I noticed there's been a significant performance degradation in our whole pipeline this morning as it was the first time we ran this patch in production.

It's still processing and no stuck tasks in "Queued", so that is very positive, but usually the pipeline is done since quite a while back, but right now it's still in the middle of processing.

**Comment by kristoffern** [comment]
Quick update on this: After I live-patched our GCP/Kubernetes cluster and the scheduler specifically to use more CPU the situation is better. We still see the warning that the scheduler does not appear to be running, just a lot less.

Normally our airflow-scheduler (release 2.2.2) pod runs on "1.5" CPU in the kubernetes cluster, and increasing it to "2" CPU last night didn't help a lot. But when I increased it to "3" this morning the situation got a lot better and the scheduler was running a bit smoother when I followed both it's log and also looking at the DAGs/tasks in the webserver. 

Good news though, no stuck tasks in "Queued" and there was a lot of ramp-up and down on workers that normally would have occurred in this case.

So this patch seems to have introduced a higher CPU load/requirement on the scheduler (In our context 1.5CPU => 3CPU).
Below is the CPU graph from our morning run and when I increased the resources to 3CPU. As you can see it still hits the limit, but overall has some more breathing room, confirmed by the systems behavior.

image

**Comment by ephraimbuddy** [comment]
cc: @ash @jedcunningham

**Comment by ephraimbuddy** [comment]
Increase the configuration to maybe every 1hr, or daily. I think that it runs too often and that's what's causing the spike:
`AIRFLOW__CELERY__STUCK_QUEUED_TASK_CHECK_INTERVAL=86400`

**Comment by kristoffern** [comment]
@ephraimbuddy Thanks I'll try it out and let you know. It currently defaults to every 5min right?
Is the operation that expensive that it takes +5min when the system is under load?

**Comment by ephraimbuddy** [comment]
I do not think the operation is expensive, and I'm not sure if this caused the spike in CPU usage but let's observe it with the above configuration

**Comment by jedcunningham** [comment]
@kristoffern, can you try with the optimization from #21335 to see if that resolves the cpu usage issue? Thanks

**Comment by kristoffern** [comment]
@jedcunningham Gladly, I'll see if I can deploy it tomorrow and how it behaves.
Just to clarify, this would be patching the system with both 19769 and 21335?

**Comment by jedcunningham** [comment]
Yep, exactly!

**Comment by kristoffern** [comment]
@ephraimbuddy and rest of the Airflow team monitoring this ticket.
I'm about to test the two patches and will get back with results later today.

But I also need to update this ticket with some other information as well.
Last week we deployed the patch for 19769 and on Friday and Saturday morning the system/pipeline was incredibly delayed due to the scheduler running at a fraction of the normal speed. We increased the CPU allocation for the scheduler as mentioned and the pipeline finished in a better shape, but it still wasn't as fast as a "clean" Airflow 2.2.3 installation.
I had also changed the settings to 1800 & 3600 seconds for clearing the Celery tasks as opposed to the standard 300 seconds.

I left the system running with the patch in place and increased CPU allocation, and when I checked in on the system on Sunday morning all was calm... too calm it turned out as nothing was being scheduled by the scheduler. DAGs would go "green" in running, but no tasks whatsoever started. Of course this meant no errors since nothing ran. 

In the end I had to redeploy our system back to a clean 2.2.3 installation and then the scheduler started working again. Of course with a few tasks that ended up in "Queued" state, but compared to the other mornings, nothing major.

The next part is only speculation, but the scheduler felt like it got slower and slower the longer it ran. With the increased time in between runs it took longer for the problem to surface, but like I said, on Sunday morning the scheduler was ground to a halt. 
Note also that the Airflow webserver **didn't** complain about the scheduler not being online, it could apparently communicate with the scheduler just fine. It was just that the scheduler didn't start any new tasks.

Again I'll test the two patches today in our staging environment and get back with the results. I just thought it was important to get the information from this weekend to you as well.

**Comment by kristoffern** [comment]
@ephraimbuddy @jedcunningham I tested by adding the second patch as well into the deployment.

It deployed without a problem, and performance-wise it seems fine at the moment (I've run a full pipeline of our ETL), but since the incident this weekend occurred after +24h I'm going to leave it running and see how it is tomorrow, so I really can't say at the moment any more than that.

However there is a problem since I now have two tasks stuck in "Queued" for close to 3hours, just like the symptoms before the fix in this PR. Before applying the second patch I didn't see that problem with 19769 only, but I could of course just gotten "lucky" and the problem might just have become much more rare. Regardless it occurred now on the first attempt with path 19769 and 21335.

image

**Comment by kristoffern** [comment]
And finally I should just mention that the Airflow webserver UI feels... sluggish and slower.
That particular tab in browser has now twice had this error (no other tabs/websites in my browser, only the Airflow webserver tab).

It's crashing with a SIGILL error and the page has to be reloaded.
image

**Comment by ephraimbuddy** [comment]
That means that the PR didn't fix the issue then cc: @jedcunningham

**Comment by jedcunningham** [comment]
Based on the issues @kristoffern had, I'm going to kick this to 2.3.0 so we have more time to properly fix this.

**Comment by ephraimbuddy** [comment]
@kristoffern I have made an adjustment to this change, see https://github.com/apache/airflow/pull/21556, apply [the new changes](https://github.com/apache/airflow/pull/21556) and test?

**Comment by ephraimbuddy** [comment]
> @kristoffern I have made an adjustment to this change, see #21556, apply [the new changes](https://github.com/apache/airflow/pull/21556) and test?

Hi @kristoffern, any update on this?

**Comment by kristoffern** [comment]
@ephraimbuddy I'll see if I can find some time today to test it. Just to double check, would that only be the #21556 patch then that should be applied, or should it be together with 

1: Patch 19769
2: Patch 21556

**Comment by kristoffern** [comment]
@ephraimbuddy I deployed to our staging cluster with the following patches applied in this order:

1: Patch 19769
2: Patch 21335
3: Patch 21556

Our pipeline is running, but I've seen at least once this warning in the webserver GUI.
image

And after ~45min of running these errors started showing in the log output for the scheduler:
image

**Comment by kristoffern** [comment]
Final note. Our pipeline finished without any tasks stuck in queued in the end.

Mid-way through I increased the CPU allocation for the scheduler from 1.5 => 3, but the AirflowTaskTimeout error kept showing up throughout the run. So it's not due to resource starvation on the scheduler end.

**Comment by ephraimbuddy** [comment]
Thanks for the update, the error is fine and can be silenced because it's not much of importance. 
I added the timeout to avoid maxing out the CPUs.
Could you leave it to run for some days? I will update the PR to silence the error

**Comment by ephraimbuddy** [comment]
> 1: Patch 19769
> 2: Patch 21335
> 3: Patch 21556

That's right. Monitor how often the webserver brings up scheduler does not appear to be running message.

**Comment by kristoffern** [comment]
@ephraimbuddy Glad to report that the scheduler and system is still responsive +24h after the deployment.
I've ran our full pipeline yesterday and now today without any hiccups.

Looks like... all good... knock on wood...

**Comment by ephraimbuddy** [comment]
Thanks @kristoffern. Please can you update with the recent changes in #21556

**Comment by ephraimbuddy** [comment]
> Thanks @kristoffern. Please can you update with the recent changes in #21556

Any update?

**Comment by kristoffern** [comment]
@ephraimbuddy Sorry, but I'm on limited amount of time currently for this and won't have any bandwidth to test it.
Again, my apologies.

**Comment by ephraimbuddy** [comment]
@Chris7 can you help and test the 3 patches here?
1: Patch #19769
2: Patch #21335
3: Patch #21556

Otherwise, we may have to revert this change

File: airflow/executors/celery_executor.py [hunk 14]
```diff
@@ -337,6 +343,8 @@ def sync(self) -> None:
 
         if self.adopted_task_timeouts:
             self._check_for_stalled_adopted_tasks()
+        if time.time() - self.stuck_tasks_last_check_time > self.stuck_queued_task_check_interval:
```
Comment (Chris7) [inline airflow/executors/celery_executor.py:346]: Won't this keep getting called every sync cycle? Shouldn't `stuck_tasks_last_check_time` be updated to the current time here.

**Comment by Chris7** [comment]
Is there an issue you can link with respect to why it needs a reversion? I do recall having some issue w/ tasks being submitted to celery during a deployment and getting left in queued limbo, but unsure what about the current code is problematic.

File: airflow/executors/celery_executor.py (hunk 14 above)
Comment (ephraimbuddy) [reply airflow/executors/celery_executor.py:346]: Good call

**Comment by ephraimbuddy** [comment]
Since you are no longer having this issue, we should revert. I called your attention because of this issue: https://github.com/apache/airflow/issues/21225 which is related to this

**Comment by ephraimbuddy** [comment]
@pingzh, Since you also experience this, could you help in testing it? We have these patches 
1: Patch https://github.com/apache/airflow/pull/19769
2: Patch https://github.com/apache/airflow/pull/21335
3: Patch https://github.com/apache/airflow/pull/21556
and on the verge of reverting all if no one tests them

**Comment by pingzh** [comment]
@ephraimbuddy since we switched to SQS as our message broker from redis (it only happened when under high load) 2 years, we haven't experienced this issue.

**Comment by repl-chris** [comment]
@ephraimbuddy we can reproduce this reliably in an isolated test environment, and would love to help test this fix to help get it merged up.  We're running with a redis broker and pgsql results db, on k8s with KEDA auto scaling - it happens when under load during a scale-in event (shutting down a worker)

We've spent a bit of time tracking it down and (at least in our case) it looks to be a problem in celery (possibly https://github.com/celery/celery/issues/7266). Airflow throws the task at celery and it just never executes, never makes it into the `celery_taskmeta` table. Airflow gets a celery task_id back and assumes its running, polling celery for updated status returns `celery_states.PENDING` indefinitely, as it has vanished and that's celery's default response for an unknown task. It seems like celery's warm shutdown on SIGTERM somehow loses tasks sometimes. We've run some tests with [this patch](https://paste.ee/p/YdNAl) and it seems to recover from this state nicely - I believe it's (almost exactly) what you had implemented in [pull 21556](https://github.com/apache/airflow/pull/21556).  

I do have some concerns around `self.task` - in your implementation when setting the task back to `SCHEDULED` it is never cleared out of `self.task`, so if the re-scheduled task happens to be picked up by a different scheduler instance it will just stay in there forever (and in `self.running` come to think of it - they'd potentially accumulate over time and eat up all the open slots). I think it would be desirable to add a filter so each executor will only "re-schedule" tasks which it "owns" - this way the executor can clean up its own internal structures at the same time....and any tasks which are not owned by a running executor will get picked up by the adoption code and then "re-scheduled" by the new owner.  I tried taking a stab at that change but it didn't quite work and I haven't figured out why yet....so maybe there's something going on I don't fully understand, I'm going to keep poking around at it though.  Let me know if I should open a new PR, or if you'd like to resurrect your existing one...

**Comment by ephraimbuddy** [comment]
Hi @repl-chris , thanks for the detailed description. 
I think it'd be best if you work on this since you can reproduce it. That way we won't go back and forth on testing.
So please go ahead with a new PR.
